File: packages/core/src/main/ts/buildstamp.ts
import * as cp from 'node:child_process'
import * as fs from 'node:fs/promises'
import * as path from 'node:path'
import * as process from 'node:process'

import {
  IBuildstamp,
  IBuildstampOptions,
  IBuildstampOptionsNormalized,
  IGitInfo
} from './interface'

export const normalizeOpts = ({
  cwd = process.cwd(),
  output = 'buildstamp.json',
  ci = true,
  git = true,
  date = true,
  extra = {}
}: IBuildstampOptions = {}): IBuildstampOptionsNormalized => ({
  ci,
  cwd,
  output,
  git,
  date,
  extra,
})

export const buildstamp = async (opts?: IBuildstampOptions): Promise<IBuildstamp> => {
  const {ci, git, date, cwd, output, extra} = normalizeOpts(opts)
  const stamp: IBuildstamp = {...extra}

  if (date) {
    stamp.date = new Date().toISOString()
  }
  if (git) {
    Object.assign(stamp, await getGitInfo(cwd, process.env))
  }
  if (ci) {
    Object.assign(stamp, getCIInfo(process.env))
  }
  if (output) {
    const file = path.resolve(cwd, output)
    await fs.mkdir(path.dirname(file), { recursive: true })
    await fs.writeFile(file, JSON.stringify(stamp, null, 2))
  }

  return stamp
}

export const getCommitBranch = async (cwd: string, env: Record<string, string | undefined>) =>
  env.CI_COMMIT_BRANCH || env.GITHUB_REF_NAME || (await spawn('git', ['rev-parse', '--abbrev-ref', 'HEAD'], cwd)).stdout

export const getGitInfo = async (cwd: string, env: Record<string, string | undefined>): Promise<IGitInfo> => {
  const { stdout: git_commit_id } = await spawn('git', ['rev-parse', 'HEAD'], cwd)
  const git_repo_url = await getGitRepoUrl(cwd)
  const git_commit_branch = await getCommitBranch(cwd, env)
  const git_repo_name = (git_repo_url.match(/([^./:]+\/[^./]+)(\.git)?$/) || [])[1]

  return {
    git_commit_branch,
    git_commit_id,
    git_repo_url,
    git_repo_name
  }
}

const getGitRepoUrl = async (cwd: string): Promise<string> => {
  const { stdout: git_list } = await spawn('git', ['remote', 'show'], cwd)
  const { stdout: git_repo_url } = await spawn('git', ['config', '--get', `remote.${git_list}.url`], cwd)
  return git_repo_url
}

// https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables
// https://docs.gitlab.com/ee/ci/variables/predefined_variables.html
// https://www.jetbrains.com/help/teamcity/predefined-build-parameters.html#Predefined+Server+Build+Parameters
export const getCIInfo = (env: Record<string, string | undefined>) => ({
  ci_run_id: env.BUILD_NUMBER || env.CI_JOB_ID || env.GITHUB_RUN_ID,
  ci_run_url: env.BUILD_URL || env.CI_JOB_URL || (env.GITHUB_RUN_ID && `${env.GITHUB_SERVER_URL}/${env.GITHUB_REPOSITORY}/actions/runs/${env.GITHUB_RUN_ID}`)
})

export const spawn = (
  cmd: string,
  args: ReadonlyArray<string>,
  cwd: string
): Promise<{stdout: string, stderr: string, status: number | null, signalCode: string | null, duration: number}> => new Promise((resolve, reject) => {
  let status: number | null = 0
  const now = Date.now()
  const stderr: string[] = []
  const stdout: string[] = []
  const p = cp.spawn(cmd, args, {cwd})

  p.stdout.on('data', (data) => stdout.push(data.toString()))
  p.stderr.on('data', (data) => stderr.push(data.toString()))

  p
    .on('error', (e) => stderr.push(e.toString()))
    .on('exit', (code) => { status = code })
    .on('close', () => {
      (status ? reject : resolve)({
        stderr: stderr.join('').trim(),
        stdout: stdout.join('').trim(),
        status,
        signalCode: p.signalCode,
        duration: Date.now() - now,
      })
    })
})

File: packages/bin/src/main/go/buildstamp/buildstamp.go
package buildstamp

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

func GetBuildstamp (opts BuildstampOpts) string {
	var (
		date    string
		gitInfo GitInfo
		ciInfo  CIInfo
		extra   map[string]string
	)

	if opts.Date {
		date = time.Now().Format(time.RFC3339)
	}

	if opts.Git {
		gitInfo = GetGitInfo(opts.Cwd)
	}

	if opts.Ci {
		ciInfo = GetCIInfo()
	}

	if err := json.Unmarshal([]byte(opts.Extra), &extra); err != nil {
		panic(err)
	}

	var buildstamp, _ = json.MarshalIndent(Buildstamp{
		date,
		gitInfo.CommitId,
		gitInfo.RepoUrl,
		gitInfo.CommitId,
		gitInfo.RepoName,
		ciInfo.RunId,
		ciInfo.RunUrl,
	}, "", "  ")

	return Mixin(string(buildstamp[:]), extra)
}

func Cwd() string {
	//if cwd, exists := os.LookupEnv("BUILDSTAMP_CWD"); exists {
	//	if abs, err := filepath.Abs(cwd); err == nil {
	//		return abs
	//	}
	//}

	ex, err := os.Getwd()
	if err != nil {
		panic(err)
	}
	return ex
}

func GetPkgInfo(cwd string) PackageJson {
	var file, _ = os.ReadFile(filepath.Join(cwd, "./package.json"))
	var data = PackageJson{}

	_ = json.Unmarshal([]byte(file), &data)

	return data
}

func GetGitInfo(cwd string) GitInfo {
	var _, commitId, _ = invoke("git", []string{"rev-parse", "HEAD"}, cwd)
	var _, repoUrl, _ = invoke("git", []string{"config", "--get", "remote.origin.url"}, cwd)
	var repoName = string(repoUrl[strings.LastIndex(repoUrl, ":")+1 : strings.LastIndex(repoUrl, ".")])
	var commitBranch = getCommitBranch(cwd)

	return GitInfo{
		commitId,
		commitBranch,
		repoUrl,
		repoName,
	}
}

func getCommitBranch(cwd string) string {
	var commitBranch = getFirstNonEmpty(os.Getenv("CI_COMMIT_BRANCH"), os.Getenv("GITHUB_REF_NAME"))
	if commitBranch == "" {
		var _, branch, _ = invoke("git", []string{"rev-parse", "--abbrev-ref", "HEAD"}, cwd)
		commitBranch = branch
	}

	return commitBranch
}

func GetCIInfo() CIInfo {
	var runId = getFirstNonEmpty(os.Getenv("BUILD_NUMBER"), os.Getenv("CI_JOB_ID")+os.Getenv("GITHUB_RUN_ID"))
	var runUrl = getFirstNonEmpty(os.Getenv("BUILD_URL"), os.Getenv("GITHUB_RUN_ID"))

	if os.Getenv("GITHUB_RUN_ID") == "" {
		runUrl = os.Getenv("GITHUB_SERVER_URL") + "/" + os.Getenv("GITHUB_REPOSITORY") + "/actions/runs/" + os.Getenv("GITHUB_RUN_ID")
	}

	return CIInfo{runId, runUrl}
}

func getFirstNonEmpty(args ...string) string {
	for _, elem := range args {
		if elem != "" {
			return elem
		}
	}
	return ""
}

func invoke(cmd string, args []string, dir string) (int, string, string) {
	var outb, errb bytes.Buffer
	var code int = 0

	cp := exec.Command(cmd, args...)
	cp.Dir = dir

	cp.Stdout = &outb
	cp.Stderr = &errb

	if errors.Is(cp.Err, exec.ErrDot) {
		cp.Err = nil
	}

	if err := cp.Run(); err != nil {
		if exitError, ok := err.(*exec.ExitError); ok {
			code = exitError.ExitCode()
		}
	}

	return code, strings.Trim(outb.String(), "\n"), strings.Trim(errb.String(), "\n")
}

func Mixin(target string, data map[string]string) string {
	result := strings.TrimRight(target, ",\n}")
	for k := range data {
		result += fmt.Sprintf(",\n  \"%s\": \"%s\"", k, data[k])
	}

	return result + "\n}"
}
